### PR TITLE
Opencl Bernoulli lpmf

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Implements `bernoulli_lpmf` using OpenCL, enabling it to use GPUs. 

This PR includes lots of changed files, as I had some problems with circular includes, fixing which involved moving `is_kernel_expression.hpp` to `prim/meta`.

Also includes a bugfix for `check_cl`, which caused the generated kernel not to compile if `check_cl` is used in a kernel that requires specific local size (any kernel that uses a colwise reduction).

## Tests

Tests check that OpenCL overload produces same results as Eigen one.

## Side Effects

None.

## Release notes

Implemented `bernoulli_lpmf` using OpenCL, enabling it to use GPUs. Fixed a bug that caused a generated kernel not to compile if it used both `check_cl` and any colwise reduction.

## Checklist

- [ ] Math issue #2152 

- [ ] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

**Comment by t4c1** [comment]
Benchmarks! 

First is the usual comparison of OpenCL version running on GPU with Eigen implementation. However @wds15 asked me how these compare when run on CPU, so I included these too. I tried to make it fair to Eigen implementation by adding another version that also runs in parallel on CPU using reduce_sum. However this turned out to be slow. @wds15 this is the first time I am using `reduce_sum`. Can you check that I did not do something stupid? The benchmark code is here: https://gist.github.com/t4c1/81975eec4f90707e8a727884f901249b

The results are a bit noisy as I ran the benchmark only once, but it is clear that for large data sizes this is the fastest implementation even when running on a CPU.

bernoulli_lpmf_cl

**Comment by t4c1** [comment]
@wds15 Did you miss this mention? Or are you too busy with other things?

**Comment by t4c1** [comment]
I fixed a bug that caused the OpenCL implementation to be slow. The graph above is updated with new values.

**Comment by wds15** [comment]
Missed it... will take a look . But Bernoulli is way too memory bound for reduce sum...so it’s expected to behave badly. How many cpus did you use for the reduce sum run?

**Comment by t4c1** [comment]
I have one 4-core CPU (i5 2500).

**Comment by t4c1** [comment]
Also Bernoulli needs to calculate log and log1m of theta, so I am not sure if it is memory bound. GPU speedup is also larger than what I would expect from memory bound operation.

**Comment by wds15** [comment]
That’s still too cheap to offset the overhead from reduce sum. A poisson with a log gamma is better. So you used 4 cores with reduce sum?

**Comment by t4c1** [comment]
I think so. Yeah. I just checked CPU usage goes to 100%, when running reduce sum. I also noticed it needs huge amounts of RAM (16GB for the largest size).

**Comment by wds15** [comment]
So you have N data items and you also have a parameter vector of the same size as N if I see that right. Then you should not slice the data y, but rather the parameter vector theta. Otherwise it's very inefficient.

**Comment by t4c1** [comment]
As far as I see it, I need to slice both. So I let `reduce_sum` slice the `std::vector<int>` and I slice the `Eigen::VectorXd`. Or am I misunderstanding something?

**Comment by rok-cesnovar** [comment]
The difference is that entire non-sliced arguments are copied to each thread, while they only get the slices from the sliced arg. And copying entire std::vector<int> is better than entire  vars vectors.

**Comment by wds15** [comment]
You need to slice both, yes.. but if you want things to run efficient, then you let reduce_sum slice the parameters and slice the data yourself in the function.

Please also make sure that you define the environment variable `STAN_NUM_THREADS` when you run this thing to be 4.

**Comment by t4c1** [comment]
Thanks, I will try it that way.

**Comment by t4c1** [comment]
I fixed the reduce_sum benchmark (I was also passing around some things by value. Ouch. That explains the memory usage). Now the results make more sense. OpenCL implementation is still the fastest for large amount of data, but otherwise reduce_sum is faster. 

bernoulli_lpmf_cl

File: stan/math/opencl/prim/bernoulli_lpmf.hpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ * @tparam T_n_cl type of integer parameters
+ * @tparam T_prob_cl type of chance of success parameters
+ * @param n integer parameter
+ * @param theta chance of success parameter
+ * @return log probability or log sum of probabilities
+ * @throw std::domain_error if theta is not a valid probability
+ * @throw std::invalid_argument if container sizes mismatch.
+ */
+template <
+    bool propto, typename T_n_cl, typename T_prob_cl,
```
Comment (rok-cesnovar) [inline stan/math/opencl/prim/bernoulli_lpmf.hpp:29]: propto is not in doxy. But its also not listed in CPU distributions (I checked like 5 and then stopped). So I guess this is fine.

**Review by rok-cesnovar — APPROVED** [review]
LGTM.

**Pull request merged**